src/libs/API.js

```diff
@@ -105,7 +105,30 @@ function makeRequestWithSideEffects(command, apiCommandParameters = {}, onyxData
  * @param {Object} [onyxData.failureData] - Onyx instructions that will be passed to Onyx.update() when the response has jsonCode !== 200.
  */
 function read(command, apiCommandParameters, onyxData) {
-    makeRequestWithSideEffects(command, apiCommandParameters, onyxData, CONST.API_REQUEST_TYPE.READ);
+    SequentialQueue.flush().then((isQueueRunning) => {
+        // If the SequentialQueue is already running,
+        // we'll simply add this read request on to the end of the sequential queue
+        // so that it gets executed after all the write requests on the queue.
+        if (isQueueRunning) {
+            const data = {
+                ...apiCommandParameters,
+                appversion: pkg.version,
+                apiRequestType: CONST.API_REQUEST_TYPE.READ,
+            };
+
+            const request = {
+                command,
+                data: {
+                    ...data,
+                },
+                ...onyxData,
+            };
+            SequentialQueue.push(request);
+            return;
+        }
+
+        return makeRequestWithSideEffects(command, apiCommandParameters, onyxData, CONST.API_REQUEST_TYPE.READ);
+    });
```

Comment:
Chatted with @jasperhuangg in a DM and have some concerns about pushing non write requests into the sequential queue. 

And curious if something like this would work:

```js
function read(command, apiCommandParameters, onyxData) {
    SequentialQueue.getActiveRun().then(() => 
        makeRequestWithSideEffects(command, apiCommandParameters, onyxData, CONST.API_REQUEST_TYPE.READ),
    );
}
```

Then we just need to make sure the `getActiveRun()` returns a `Promise.resolve()` or the active call to `process()`.

```diff
diff --git a/src/libs/Network/SequentialQueue.js b/src/libs/Network/SequentialQueue.js
index fc670d118..07b44a386 100644
--- a/src/libs/Network/SequentialQueue.js
+++ b/src/libs/Network/SequentialQueue.js
@@ -40,6 +40,8 @@ function process() {
     return task.then(process);
 }

+let activeRun = Promise.resolve();
+
 function flush() {
     if (isSequentialQueueRunning) {
         return;
@@ -63,16 +65,21 @@ function flush() {
         key: ONYXKEYS.PERSISTED_REQUESTS,
         callback: () => {
             Onyx.disconnect(connectionID);
-            process()
+            activeRun = process()
                 .finally(() => {
                     isSequentialQueueRunning = false;
                     resolveIsReadyPromise();
                     currentRequest = null;
+                    activeRun = Promise.resolve();
                 });
         },
     });
 }

+function getActiveRun() {
+    return activeRun;
+}
+
 /**
  * @returns {Boolean}
  */
```

Or maybe it's still necessary to flush the queue - I didn't quite catch why we were doing that.

Reply:
I tried this which works: https://github.com/Expensify/App/pull/12103/files

But if what you suggested works I think that's better because my solution involves exposing `SequentialQueue.process`

Reply:
actually @marcaaron just tried what you suggested and it doesn't seem to work

Reply:
I have your changes locally and it appears that they're not having any effect. I think what might be happening is that the promise returned from `process.finally` hasn't been set by `flush` yet (since process is still happening when the `API.read` call happens), therefore the promise returned from `getActiveRun` just resolves without waiting for `process` to finish.

Why do we need to set `activeRun = Promise.resolve();` in `process.finally`? Doesn't `process.finally` already return a Promise that resolves with `undefined` because we're not returning anything from the callback in `finally`?

Reply:
FWIW, I also agree with Marc that we shouldn't be adding requests to the sequential queue that aren't supposed to be there. It's mixing up the concerns with what exactly the sequential queue is for. That's also a good reason why I don't think there should be an array of extra requests either. The sequential queue should only be doing one thing, and that is persisting and running write requests.

I think that the sequential queue needs to have one state that the outside world is aware of:

- The sequential queue is complete and it's not doing anything

Now, there are two pieces of information there, but we should combine them into one thing for the outside world:

1. The sequential queue is complete. This means that all the requests persisted to storage have been completed.
2. The sequential queue is not doing anything. This means that there were no requests persisted to storage, and there are no write requests currently being processed.

Then implement something like `SequentialQueue.waitForIdle()` that is called before every read request.

-----------
> It also requires us to read persisted requests from disk every time before running a read request 

To get around this, we can keep an in-memory counter of how many requests are in the persisted queue. That way, the persisted queue only needs to be read from disk when the app starts.

Reply:
>I have your changes locally and it appears that they're not having any effect.

Full disclosure, I didn't test any of that so it's just a high level / pseudocode idea. We would need to experiment to make it work in practice.

>the promise returned from process.finally hasn't been set by flush yet, therefore the promise returned from getActiveRun still resolves without waiting for process to finish

I'm confused by this statement. What are we trying to do specifically when this happens? It sounds like the queue is not actually running, which is expected, and we should be OK to not wait for it to finish before performing a "read".

>I remember we also discussed storing an array of requests to run after process finishes. Once then we process this array after process.finally runs. I feel like this would work, but are there any reasons why we wouldn't want to do this?

I don't really see any problems with it besides the fact that a promise callback would achieve the same thing. No need to have a separate queue for requests if each request is waiting for a promise callback. JS has its own task queue that the callbacks will go in if they are inside a `.then()`.

Reply:
Thank you guys so much for the review! I saw and agree with [Tim's notes](https://github.com/Expensify/App/pull/12103#discussion_r1008044814) about how we shouldn't resolve different values in the Promise returned in `flush`, so I'm going to be scrapping this approach.

I took another look at Marc's approach and was able to figure some more things out. 

<img width="281" alt="Screen Shot 2022-10-31 at 12 01 44 PM" src="https://user-images.githubusercontent.com/31285285/199088826-27ad8f9b-1d01-4535-aa2f-f5b708a23767.png">

This is a screenshot of the network requests made when I changed a workspace offline and came back online. We only care about `OpenWorkspaceReimburseView` (r) and `UpdateWorkspaceCustomUnit` (w).


> I'm confused by this statement. What are we trying to do specifically when this happens? It sounds like the queue is not actually running, which is expected, and we should be OK to not wait for it to finish before performing a "read".

@marcaaron I stuck some console logs in a few places and this may shed some light on what I mean.

<img width="630" alt="Screen Shot 2022-10-31 at 12 16 23 PM" src="https://user-images.githubusercontent.com/31285285/199091616-2f14d111-a465-43ec-b511-42c2af3152fd.png">

As you can see, when we call the read command, the queue is already running. The `activeRun` just hasn't been set to `process.finally()` yet. This means a call to `SequentialQueue.getActiveRun.then()` would just resolve immediately.

Reply:
@marcaaron @tgolen I think I have a working solution with promises (pending more intense testing)! Going to push it to https://github.com/Expensify/App/pull/12219/files and request you guys for review.

I ended up going with the promise approach, but went with something similar to [the pre-existing `isReadyPromise`](https://github.com/Expensify/App/blob/826e2eaa9ccf06e61577c069213a6f421a86ad54/src/libs/Network/SequentialQueue.js#L9-L12), i.e. the `isDonePromise`.

This promise resolves when every request on the SequentialQueue has been fired AND has responded. The reason I had to do this was because tiny read requests that respond quickly could sometimes read data on the server before a write request has had a chance to modify that data (even if the read request was technically fired afterwards).

Reply:
>I stuck some console logs in a few places and this may shed some light on what I mean.

Hmm tbh I don't really know where we're logging or which code we logged in for that example so not totally sure what to take away from it. I see something about "retrying" in the logs. Are we retrying these requests? 😕 

>when we call the read command, the queue is already running. The activeRun just hasn't been set to process.finally() yet.

Well, I would say that the queue is not actually running if `process()` has not been called. So we should be good to run the read request. If there's a bug after that then we are allowing them to happen at the wrong time.

Reply:
It feels like what we need to wait for is the Onyx connection -> https://github.com/Expensify/App/blob/f656f08e0a5a88fae209dc7bbef5ef74ca8fa894/src/libs/Network/SequentialQueue.js#L61-L64

Because the queue is "running" we just haven't called `process()` yet ?

Reply:
@marcaaron

> Because the queue is "running" we just haven't called process() yet ?

That's a good point, I didn't actually try that yet before I went ahead and started that other PR. I can give it a go and see if it works.